index.js

```diff
@@ -46,184 +45,254 @@ const version = () => {
+function rc (ctx, path) {
+  if (argv.use) {
+    map()
+    return Promise.resolve()
+  }
 
-  let options = Object.assign(
-    {
-      from: filename,
-      to: output || (argv.replace ? filename : path.join(dir, path.basename(filename)))
-    },
-    config.options
-  )
-  options.to = path.resolve(options.to)
-
-  return postcss(config.plugins).process(css, options)
-    .then(result => {
-      if (result.messages.some(msg => msg.type === 'warning')) spinner.fail()
-
-      var tasks = [fs.outputFile(options.to, result.css)]
-      if (result.map) {
-        tasks.push(fs.outputFile(options.to.replace('.css', '.css.map'), result.map))
-      }
-      return Promise.all(tasks)
-        .then(() => {
-          spinner.succeed()
-          return result
-        })
+  return postcssrc(ctx, path)
+    .then((rc) => {
+      config = rc
+      map()
+    })
+    .catch((err) => {
+      if (err.message.indexOf('No PostCSS Config found') === -1) throw err
     })
 }
 
-function processFiles (files) {
-  if (typeof files === 'string') files = [files]
-  return Promise.all(files.map(file => {
+function map () {
+  if (argv.map === false) config.options.map = false
+  else if (argv.map) config.options.map = { inline: false }
+  else config.options.map = true
+}
+
+function files (files) {
+  if (typeof files === 'string') files = [ files ]
+
+  return Promise.all(files.map((file) => {
     // Read from stdin
     if (file === '-') {
-      return getStdin()
-      .then(css => processCSS(css))
+      return stdin()
+        .then((content) => css(content, 'stdin'))
     }
+
     return fs.readFile(file)
-    .then(css => processCSS(css, file))
+      .then((content) => css(content, file))
   }))
 }
 
-function getConfig (ctx, path) {
-  if (argv.use || argv.parser || argv.stringifier || argv.syntax) {
-    config = {
-      plugins: argv.use ? argv.use.map(plugin => require(plugin)) : [],
-      options: {
-        parser: argv.parser ? require(argv.parser) : undefined,
-        syntax: argv.syntax ? require(argv.syntax) : undefined,
-        stringifier: argv.stringifier ? require(argv.stringifier) : undefined
-      }
+function css (css, file) {
+  const ctx = { options: argv.ctx ? argv.ctx : {} }
+
+  if (file !== 'stdin') {
+    ctx.file = {
+      dirname: path.dirname(file),
+      extname: path.extname(file),
+      basename: path.basename(file)
     }
-    setMap()
-    return Promise.resolve()
-  } else {
-    return postcssrc(ctx, path)
-    .then(conf => {
-      config = conf
-      setMap()
-    })
-    .catch(err => {
-      if (err.message.indexOf('No PostCSS Config found') === -1) throw err
-    })
+
+    if (!argv.config) argv.config = ctx.file.dirname
   }
+
+  if (!argv.config) argv.config = process.cwd()
+
+  const time = process.hrtime()
+
+  spinner.text = `Processing ${file}`
+  spinner.start()
+
+  rc(ctx, argv.config)
+    .then(() => {
+      if (file !== 'stdin') {
+        config.options = Object.assign(
```

Comment:
You shouldn't be mutating the global `config.options`; this will create some weird timing/concurrency issues. Make a local variable containing the per-file config options.